homeassistant/components/upb/config_flow.py

```diff
@@ -0,0 +1,140 @@
+"""Config flow for UPB PIM integration."""
+import asyncio
+import logging
+from urllib.parse import urlparse
+
+import async_timeout
+import upb_lib
+import voluptuous as vol
+
+from homeassistant import config_entries, exceptions
+from homeassistant.const import CONF_ADDRESS, CONF_FILE_PATH, CONF_HOST, CONF_PROTOCOL
+
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+PROTOCOL_MAP = {"TCP": "tcp://", "Serial port": "serial://"}
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PROTOCOL, default="Serial port"): vol.In(
+            ["TCP", "Serial port"]
+        ),
+        vol.Required(CONF_ADDRESS): str,
+        vol.Required(CONF_FILE_PATH, default=""): str,
+    }
+)
+VALIDATE_TIMEOUT = 15
+
+
+async def _validate_input(data):
+    """Validate the user input allows us to connect."""
+
+    def _connected_callback():
+        connected_event.set()
+
+    connected_event = asyncio.Event()
+    file_path = data.get(CONF_FILE_PATH)
+    url = _make_url_from_data(data)
+
+    upb = upb_lib.UpbPim({"url": url, "UPStartExportFile": file_path})
+    if not upb.config_ok:
+        _LOGGER.error("Missing or invalid UPB file: %s", file_path)
+        raise InvalidUpbFile
+
+    upb.connect(_connected_callback)
+
+    try:
+        with async_timeout.timeout(VALIDATE_TIMEOUT):
+            await connected_event.wait()
+    except asyncio.TimeoutError:
+        pass
+
+    upb.disconnect()
+
+    if not connected_event.is_set():
+        _LOGGER.error(
+            "Timed out after %d seconds trying to connect with UPB PIM at %s",
+            VALIDATE_TIMEOUT,
+            url,
+        )
+        raise CannotConnect
+
+    # Return info that you want to store in the config entry.
+    return (upb.network_id, {"title": "UPB", CONF_HOST: url, CONF_FILE_PATH: file_path})
+
+
+def _make_url_from_data(data):
+    host = data.get(CONF_HOST)
+    if host:
+        return host
+
+    protocol = PROTOCOL_MAP[data[CONF_PROTOCOL]]
+    address = data[CONF_ADDRESS]
+    return f"{protocol}{address}"
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for UPB PIM."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the UPB config flow."""
+        self.importing = False
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                if self._url_already_configured(_make_url_from_data(user_input)):
+                    return self.async_abort(reason="address_already_configured")
```

Comment:
If this can't throw an exception please move it above the `try` block

Reply:
It can. At least this test can cause an exception. https://github.com/gwww/core/blob/upb-light/tests/components/upb/test_config_flow.py#L138-L152

Likelihood of it happening are small but could happen. I could write to `_make_url_from_data` to be more resilient or I can intentionally catch the exception.

Reply:
It looks like the user would have to modify the data being posted from the config flow form in order to inject invalid data to cause a `KeyError` in `_make_url_from_data`.  Could go either way on this one.